pydantic/_internal/_core_utils.py

```diff
@@ -100,51 +100,25 @@ def get_type_ref(type_: type[Any], args_override: tuple[type[Any], ...] | None =
     return type_ref
 
 
+def get_ref(s: core_schema.CoreSchema) -> None | str:
+    """Get the ref from the schema if it has one.
+    This exists just for type checking to work correctly.
+    """
+    return s.get('ref', None)  # type: ignore
+
+
 def collect_definitions(schema: core_schema.CoreSchema) -> dict[str, core_schema.CoreSchema]:
-    # Only collect valid definitions. This is equivalent to collecting all definitions for "valid" schemas,
-    # but allows us to reuse this logic while removing "invalid" definitions
-    valid_definitions = dict()
+    defs: dict[str, CoreSchema] = {}
 
     def _record_valid_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
-        ref: str | None = s.get('ref')  # type: ignore[assignment]
+        ref = get_ref(s)
         if ref:
-            metadata = s.get('metadata')
-            definition_is_invalid = isinstance(metadata, dict) and 'invalid' in metadata
-            if not definition_is_invalid:
-                valid_definitions[ref] = s
+            defs[ref] = s
         return recurse(s, _record_valid_refs)
 
     walk_core_schema(schema, _record_valid_refs)
 
-    return valid_definitions
-
-
-def remove_unnecessary_invalid_definitions(schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
```

Comment:
I have a hunch that with a little refactoring we can get rid of some of the other functions in this module, but this was the only one I was able to delete without any changes unrelated to this PR.